yarn-project/merkle-tree/src/snapshots/append_only_snapshot.ts

```diff
@@ -0,0 +1,179 @@
+import { Hasher, SiblingPath } from '@aztec/types';
+
+import { LevelUp } from 'levelup';
+
+import { AppendOnlyTree } from '../interfaces/append_only_tree.js';
+import { SiblingPathSource } from '../interfaces/merkle_tree.js';
+import { TreeBase } from '../tree_base.js';
+import { SnapshotBuilder } from './snapshot_builder.js';
+
+const nodeVersionKey = (name: string, level: number, index: bigint) =>
+  `snapshot:${name}:node:${level}:${index}:version`;
+const nodePreviousValueKey = (name: string, level: number, index: bigint) =>
+  `snapshot:${name}:node:${level}:${index}:value`;
+const snapshotMetaKey = (name: string, version: number) => `snapshot:${name}:${version}`;
+
+/**
+ * A more space-efficient way of storing snapshots of AppendOnlyTrees that trades space need for slower
+ * sibling path reads.
+ *
+ * Complexity:
+ *
+ * N - count of non-zero nodes in tree
+ * M - count of snapshots
+ * H - tree height
+ *
+ * Space complexity: O(N) (stores the previous value for each node and at which snapshot it was last modified)
+ * Sibling path access:
+ *  Best case: O(H) database reads + O(1) hashes
+ *  Worst case: O(H) database reads + O(H) hashes
+ */
+export class AppendOnlySnapshotBuilder implements SnapshotBuilder {
+  constructor(private db: LevelUp, private tree: TreeBase & AppendOnlyTree, private hasher: Hasher) {}
+  async getSnapshot(version: number): Promise<SiblingPathSource> {
+    const filledLeavesAtVersion = await this.#getLeafCountAtVersion(version);
+
+    if (typeof filledLeavesAtVersion === 'undefined') {
+      throw new Error(`Version ${version} does not exist for tree ${this.tree.getName()}`);
+    }
+
+    return new AppendOnlySnapshot(this.db, version, filledLeavesAtVersion, this.tree, this.hasher);
+  }
+
+  async snapshot(version: number): Promise<SiblingPathSource> {
+    const leafCountAtVersion = await this.#getLeafCountAtVersion(version);
+    if (typeof leafCountAtVersion !== 'undefined') {
+      throw new Error(`Version ${version} of tree ${this.tree.getName()} already exists`);
+    }
+
+    const batch = this.db.batch();
+    const root = this.tree.getRoot(false);
+    const depth = this.tree.getDepth();
+    const treeName = this.tree.getName();
+    const queue: [Buffer, number, bigint][] = [[root, 0, 0n]];
+
+    // walk the BF and update latest values
```

Comment:
What is the `BF` in this case, breadth first?

Reply:
Yes